doc/sfp-cmis/cmis-init.md

```diff
@@ -0,0 +1,216 @@
+
+# Requirement Overview
+
+This document describes functional behavior of the CMIS application initialization support in SONiC.
+
+The Common Management Interface Specification (CMIS) provides a variety of features
+and support for different transceiver form factors. A 
+CMIS transceiver may support multiple application, and the application initialization sequence is now mandatory for the dynamic port breakout mode to correctly update the active CMIS application based on the desired port mode. Otherwise the link will be down if the host port mode does not match the selected application on the CMIS transceiver.
+
+The feature is built on top of SONiC **sfp-refactor** framework to provide a platform-independent
+solution, and the individual platforms could easily enable this feaure by having its **Sfp** inherited from **SfpOptoeBase**.
+
+**Example:**  
+```
+from sonic_platform_base.sonic_xcvr.sfp_optoe_base import SfpOptoeBase
+
+class Sfp(SfpOptoeBase):
+
+    def __init__(self, sfp_index):
+        SfpOptoeBase.__init__(self)
+        self.index = sfp_index
+
+    def get_port_type(self):
+        return self.SFP_PORT_TYPE_QSFPDD
+
+    def get_eeprom_path(self):
+        # platform-specific per-port I2C bus
+        i2c_bus = 32 + self.index
+        return "/sys/bus/i2c/devices/{}-0050/eeprom".format(i2c_bus)
+```
+
+The scope of this feature is as follow:
+
+- **sonic-platform-common**  
+  - Enhance the **sonic-xcvr** to support CMIS application advertising and initialization
+- **sonic-platform-daemons**  
+  - Enhance the **sonic-xcvrd** to support paralleled CMIS application initialization
+- **sonic-utilities**  
+  - Enhance the **sfputil** and **sfpshow** to show CMIS application advertisement
+
+## Functional Requirements
+
+1. Ability to parse the advertised applications from the transceivers
+2. Ability to post the advertised applications to the STATE_DB
+3. Ability to support parallel processing of the application initialization
+4. Ability to detect the errors of the application initialization on the transceivers
+
+**Note:**  
+The duration of the CMIS application initialization greatly differs from transceivers
+to transceivers, while some take 3 seconds for activating the 4x100G mode, some require 15 seconds.
+Hence it's mandatory to support parallel processing.
+
+## Warm Boot Requirements
+
+Functionality should continue to work across warm boot. 
+- The CMIS application initialization should not be performed during WarmBoot.
+- The CMIS application initialization should be skipped if no application code updates.
+
+# Functional Description
+
+- The **pmon#xcvrd** will detect the module type of the attched transceivers and post the
+information to the **STATE_DB**
+- If the transceiver is a CMIS module, the **pmon#xcvrd** will activate the appropriate
+application mode as per the current port mode.
```

Comment:
As I have commented before, I would suggest adding a knob here,  not all the platforms need to config this from the NOS layer.  please refer to the media setting function currently supported by xcvrd,  it provides a configuration file, and only the platform which needs it will turn it on via setting the configuration file.

Reply:
Yep, you're right, maybe we should enhance the media_setting.json to conditionally activate the application initialization

Reply:
> Is the suggestion to have a per-vendor/MPN setting in media_settins or a "global" setting? I'd be fairly strongly against a per-module setting since it means we must have advance knowledge that the module will be inserted in order to do CMIS-compliant initialization, or a "default" setting for it.

I would suggest a "global" setting, which allows a vendor to disable(not start) this new function on its platform(s).

Reply:
As per the last sync up with Prince, if I'm right about it, we'll drop these exception handlers from media_settings.json until there is a true requirement/request for this. Hence the references to media_settings.json have been removed in the latest code PRs.

Reply:
> As per the last sync up with Prince, if I'm right about it, we'll drop these exception handlers from media_settings.json until there is a true requirement/request for this. Hence the references to media_settings.json have been removed in the latest code PRs.

then how to address the requirement that some platforms may not need to activate this function?

Reply:
Sorry, but I'm not following you, the CMIS init is a feature of the transceiver, it's not platform-specific, why do we want to have it disabled on some certain platform?

Reply:
> Sorry, but I'm not following you, the CMIS init is a feature of the transceiver, it's not platform-specific, why do we want to have it disabled on some certain platform?

a possibility is that some vendors have implemented this function in a lower layer,  no extra configuration is needed from SONiC, so they can skip this function.

Reply:
Copy that, thanks for the clarification, will add one additional flag into the platform.json for this